include/seqan3/alphabet/union_alphabet.hpp

```diff
@@ -43,225 +43,292 @@
     {
-        return value;
+        return _value;
     }
 
     //! assign from a character
     constexpr union_alphabet & assign_char(char_type const c)
     {
-        auto sizes_it = sizes.begin();
-        auto sizes_it_end = sizes.end();
-        auto alphabets_it = alphabets.begin();
-        auto size_start = 0;
-
-        bool found = false;
-
-        while(sizes_it != sizes_it_end && !found)
-        {
-            std::visit([&](auto alphabet)
-            {
-                alphabet.assign_char(c);
-                if (alphabet.to_char() == c)
-                {
-                    value = size_start + alphabet.to_rank();
-                    found = true;
-                }
-            }, *alphabets_it);
-
-            size_start += *sizes_it;
-            sizes_it++;
-            alphabets_it++;
-        }
-
-        if (!found)
-        {
-            value = 0;
-        }
-
+        _value = char_to_value[c];
         return *this;
     }
 
     //! assign from a numeric value
     constexpr union_alphabet & assign_rank(rank_type const i)
     {
         assert(i < value_size);
-        value = i;
+        _value = i;
         return *this;
     }
 
     //! @name comparison operators
     //!@{
     constexpr bool operator==(union_alphabet const & rhs) const
     {
-        return value == rhs.value;
+        return _value == rhs._value;
     }
 
     constexpr bool operator!=(union_alphabet const & rhs) const
     {
-        return value != rhs.value;
+        return _value != rhs._value;
     }
 
     constexpr bool operator<(union_alphabet const & rhs) const
     {
-        return value < rhs.value;
+        return _value < rhs._value;
     }
 
     constexpr bool operator>(union_alphabet const & rhs) const
     {
-        return value > rhs.value;
+        return _value > rhs._value;
     }
 
     constexpr bool operator<=(union_alphabet const & rhs) const
     {
-        return value <= rhs.value;
+        return _value <= rhs._value;
     }
 
     constexpr bool operator>=(union_alphabet const & rhs) const
     {
-        return value >= rhs.value;
+        return _value >= rhs._value;
     }
     //!@}
+
+protected:
+    //! \privatesection
+    // conversion tables
+
+    static constexpr rank_type from_base_(variant_type const & alphabet_v)
+    {
+        return std::visit([&](auto && alphabet) -> rank_type
+        {
+            return prefix_sum_sizes[alphabet_v.index()] + static_cast<rank_type>(alphabet.to_rank());
+        }, alphabet_v);
+    }
+
+    static constexpr auto prefix_sum_sizes
+        = detail::alphabet_prefix_sum_sizes<first_alphabet_type, alphabet_types...>();
+
+    static constexpr auto value_to_char
+        = detail::union_alphabet::value_to_char_table<char_type, first_alphabet_type, alphabet_types...>();
+
+    static constexpr auto char_to_value
+        = detail::union_alphabet::char_to_value_table<char_type, first_alphabet_type, alphabet_types...>();
```

Comment:
you could also move the functions from above (detail) to be private members or direct-initialize the member variables with local lambda functions, like here:
https://github.com/seqan/seqan3/blob/master/include/seqan3/alphabet/composition.hpp#L159
That would increase the locality of the code since they are obviously not used anywhere else.